File: backend/app/api/v1/payroll_config.py
"""
工资项配置 API 路由
/api/v1/payroll-config/

GET    /items          工资项列表
POST   /items          新增/更新工资项
DELETE /items/{code}   删除工资项（非系统项）
GET    /rules          薪资规则列表
PUT    /rules/{code}   更新薪资规则
POST   /init           初始化默认配置
"""
from fastapi import APIRouter, Depends, Query, Request
from pydantic import BaseModel
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.services.payroll_config import PayrollConfigService
from app.utils.deps import get_store_id, require_role, make_response

router = APIRouter()


class ItemConfigRequest(BaseModel):
    item_code: str
    item_name: str
    item_type: str  # income / deduction
    data_source: str  # contract / attendance / performance / kpi / rule / manual
    formula_ast: dict | None = None
    default_value: float = 0
    sort_order: int = 0
    note: str | None = None


class RuleUpdateRequest(BaseModel):
    rule_value: float


@router.get("/items", summary="工资项列表")
async def list_items(
    request: Request,
    db: AsyncSession = Depends(get_db),
):
    store_id = get_store_id(request)
    service = PayrollConfigService(db, store_id)
    await service.init_default_config_if_empty()
    items = await service.get_items_config()
    return make_response(
        data=[
            {
                "id": i.id,
                "item_code": i.item_code,
                "item_name": i.item_name,
                "item_type": i.item_type,
                "data_source": i.data_source,
                "formula_ast": i.formula_ast,
                "default_value": float(i.default_value or 0),
                "is_active": i.is_active,
                "is_system": i.is_system,
                "sort_order": i.sort_order,
                "note": i.note,
            }
            for i in items
        ],
        request=request,
    )


@router.post("/items", summary="新增/更新工资项(老板)")
async def upsert_item(
    request: Request,
    body: ItemConfigRequest,
    db: AsyncSession = Depends(get_db),
):
    require_role(request, ["boss"])
    store_id = get_store_id(request)
    service = PayrollConfigService(db, store_id)
    item = await service.upsert_item_config(
        item_code=body.item_code,
        item_name=body.item_name,
        item_type=body.item_type,
        data_source=body.data_source,
        formula_ast=body.formula_ast,
        sort_order=body.sort_order,
        default_value=body.default_value,
        note=body.note,
    )
    await db.commit()
    return make_response(message=f"工资项 {body.item_code} 已保存", request=request)


@router.delete("/items/{item_code}", summary="删除工资项(老板)")
async def delete_item(
    request: Request,
    item_code: str,
    db: AsyncSession = Depends(get_db),
):
    require_role(request, ["boss"])
    store_id = get_store_id(request)
    from sqlalchemy import select, and_, delete as del_stmt
    from app.models.payroll_config import PayrollItemConfig
    stmt = select(PayrollItemConfig).where(
        and_(
            PayrollItemConfig.store_id == store_id,
            PayrollItemConfig.item_code == item_code,
        )
    )
    result = await db.execute(stmt)
    item = result.scalar_one_or_none()
    if not item:
        from app.utils.exceptions import NotFoundError
        raise NotFoundError("工资项不存在")
    if item.is_system:
        from app.utils.exceptions import ValidationError
        raise ValidationError("系统内置工资项不可删除")
    await db.execute(
        del_stmt(PayrollItemConfig).where(
            and_(
                PayrollItemConfig.store_id == store_id,
                PayrollItemConfig.item_code == item_code,
            )
        )
    )
    await db.commit()
    return make_response(message=f"工资项 {item_code} 已删除", request=request)


@router.get("/rules", summary="薪资规则列表")
async def list_rules(
    request: Request,
    db: AsyncSession = Depends(get_db),
):
    store_id = get_store_id(request)
    service = PayrollConfigService(db, store_id)
    await service.init_default_config_if_empty()
    rules = await service.get_salary_rules()
    # 取完整规则对象
    from sqlalchemy import select
    from app.models.payroll_config import SalaryRule
    stmt = select(SalaryRule).where(SalaryRule.store_id == store_id)
    result = await db.execute(stmt)
    rule_objs = result.scalars().all()
    return make_response(
        data=[
            {
                "id": r.id,
                "rule_code": r.rule_code,
                "rule_name": r.rule_name,
                "rule_value": float(r.rule_value),
                "rule_unit": r.rule_unit,
                "is_active": r.is_active,
                "note": r.note,
            }
            for r in rule_objs
        ],
        request=request,
    )


@router.put("/rules/{rule_code}", summary="更新薪资规则(老板)")
async def update_rule(
    request: Request,
    rule_code: str,
    body: RuleUpdateRequest,
    db: AsyncSession = Depends(get_db),
):
    require_role(request, ["boss"])
    store_id = get_store_id(request)
    service = PayrollConfigService(db, store_id)
    rule = await service.update_salary_rule(rule_code, body.rule_value)
    if not rule:
        from app.utils.exceptions import NotFoundError
        raise NotFoundError("薪资规则不存在")
    await db.commit()
    return make_response(message=f"规则 {rule_code} 已更新", request=request)


@router.post("/init", summary="初始化默认配置(老板)")
async def init_config(
    request: Request,
    db: AsyncSession = Depends(get_db),
):
    require_role(request, ["boss"])
    store_id = get_store_id(request)
    service = PayrollConfigService(db, store_id)
    await service.init_default_config_if_empty()
    await db.commit()
    return make_response(message="默认配置已初始化", request=request)

File: frontend/src/api/payrollConfig.ts
// 工资项配置 API（公式编辑器使用）
import client from './client'
import type { ApiResponse } from './types'

// ====== 公式 AST 类型定义（与后端 formula_engine.py 对齐）======

export type FormulaNode =
  | { type: 'const'; value: number; isPlaceholder?: boolean }
  | { type: 'field'; source: FormulaSource; field: string }
  | { type: 'op'; op: '+' | '-' | '*' | '/'; left: FormulaNode; right: FormulaNode }
  | { type: 'func'; name: 'max' | 'min' | 'round' | 'abs' | 'if'; args: FormulaNode[] }

export type FormulaSource =
  | 'contract'
  | 'attendance'
  | 'performance'
  | 'kpi'
  | 'rule'
  | 'manual'

export type DataSource =
  | 'contract'
  | 'attendance'
  | 'performance'
  | 'kpi'
  | 'rule'
  | 'manual'

export type ItemType = 'income' | 'deduction'

export interface PayrollItemConfig {
  id: string
  item_code: string
  item_name: string
  item_type: ItemType
  data_source: DataSource
  formula_ast: FormulaNode | null
  default_value: number
  is_active: boolean
  is_system: boolean
  sort_order: number
  note: string | null
}

export interface SalaryRule {
  id: string
  rule_code: string
  rule_name: string
  rule_value: number
  rule_unit: string | null
  is_active: boolean
  note: string | null
}

export interface ItemConfigRequest {
  item_code: string
  item_name: string
  item_type: ItemType
  data_source: DataSource
  formula_ast: FormulaNode | null
  default_value: number
  sort_order: number
  note?: string
}

// ====== 数据源字段映射（公式编辑器使用）======

export const SOURCE_FIELDS: Record<FormulaSource, { field: string; label: string }[]> = {
  contract: [
    { field: 'monthly_salary', label: '合同月薪' },
    { field: 'base_salary', label: '底薪' },
    { field: 'meal_allowance', label: '餐补' },
    { field: 'allowance', label: '津贴' },
  ],
  attendance: [
    { field: 'late_count', label: '迟到次数' },
    { field: 'total_late_minutes', label: '迟到总分钟' },
    { field: 'absent_count', label: '旷工天数' },
    { field: 'early_count', label: '早退次数' },
    { field: 'present_days', label: '出勤天数' },
    { field: 'total_days', label: '应出勤天数' },
  ],
  performance: [
    { field: 'total_amount', label: '业绩总额' },
    { field: 'booking', label: '订桌业绩' },
    { field: 'wework_payment', label: '企微收款业绩' },
    { field: 'bottle', label: '存酒业绩' },
    { field: 'card', label: '卡券业绩' },
  ],
  kpi: [
    { field: 'coefficient', label: 'KPI系数' },
    { field: 'total_score', label: 'KPI总分' },
  ],
  rule: [
    { field: 'late_deduction_per_minute', label: '迟到每分钟扣款' },
    { field: 'absent_factor', label: '旷工扣款倍数' },
    { field: 'early_deduction_per_time', label: '早退每次扣款' },
    { field: 'commission_rate', label: '提成比例' },
    { field: 'rest_days_per_month', label: '月休息天数' },
  ],
  manual: [],
}

export const SOURCE_LABELS: Record<FormulaSource, string> = {
  contract: '合同',
  attendance: '考勤',
  performance: '业绩',
  kpi: 'KPI',
  rule: '规则',
  manual: '手工',
}

export const FUNCTIONS = [
  { name: 'max', label: '最大值 max(a,b,...)', argCount: 2 },
  { name: 'min', label: '最小值 min(a,b,...)', argCount: 2 },
  { name: 'round', label: '四舍五入 round(x,位数)', argCount: 2 },
  { name: 'abs', label: '绝对值 abs(x)', argCount: 1 },
  { name: 'if', label: '条件 if(条件,真,假)', argCount: 3 },
] as const

// ====== API 方法 ======

export function listItems() {
  return client.get<ApiResponse<PayrollItemConfig[]>>('/payroll-config/items')
}

export function upsertItem(body: ItemConfigRequest) {
  return client.post<ApiResponse<null>>('/payroll-config/items', body)
}

export function deleteItem(item_code: string) {
  return client.delete<ApiResponse<null>>(`/payroll-config/items/${item_code}`)
}

export function listRules() {
  return client.get<ApiResponse<SalaryRule[]>>('/payroll-config/rules')
}

export function updateRule(rule_code: string, rule_value: number) {
  return client.put<ApiResponse<null>>(`/payroll-config/rules/${rule_code}`, {
    rule_value,
  })
}

export function toggleRule(rule_code: string, is_active: boolean) {
  return client.put<ApiResponse<null>>(`/payroll-config/rules/${rule_code}/toggle`, {
    is_active,
  })
}

export function initConfig() {
  return client.post<ApiResponse<null>>('/payroll-config/init')
}
